### DISCUSSION TITLE
Locate PO files from an absolute path for Localization

### DISCUSSION THREAD

**Discussion by VenkatCharyBhairoju** [OrchardCMS/OrchardCore#16265] [Q&A] [answered]
To Locate PO files from an absolute path for localization, I have tried implementing OrchardCore.Localization.PortableObject.ILocalizationFileLocationProvider interface and registered as a service. But it is not working for me and I don't find any solutions/samples online. Could you please suggest me a sample code to achieve this. I am using .net8.0

**Comment by hishamco** [comment]
Please show us your registration, also have a look to PO files locations https://docs.orchardcore.net/en/latest/reference/modules/Localize/#po-files-locations

**Reply by VenkatCharyBhairoju** [reply]
Hi Hishamco, 

Thanks for your quick reply.

**Changes in Program.cs file**

// Absolute path to the localization resources
var localizationPath = "C:/inetpub/locale";

// Register the custom localization file location provider
builder.Services.AddSingleton<ILocalizationFileLocationProvider>(serviceProvider =>
{
    var logger = serviceProvider.GetRequiredService<ILogger<CustomLocalizationFileLocationProvider>>();
    logger.LogInformation("Registering CustomLocalizationFileLocationProvider");
    return new CustomLocalizationFileLocationProvider(localizationPath, logger);
});

builder.Services.AddPortableObjectLocalization(options => options.ResourcesPath = localizationPath);

// Configure request localization
builder.Services.Configure<RequestLocalizationOptions>(options =>
{
    var supportedCultures = new[] { "en", "fr", "es" };
    options.DefaultRequestCulture = new RequestCulture(supportedCultures[2]);
    options.SupportedCultures = supportedCultures.Select(c => new CultureInfo(c)).ToList();
    options.SupportedUICultures = supportedCultures.Select(c => new CultureInfo(c)).ToList();
    options.RequestCultureProviders.Insert(0, new CookieRequestCultureProvider());

    options.RequestCultureProviders.Clear();
});

**Added the below middleware**
app.UseRequestLocalization();

**CustomLocalizationFileLocationProvider Class** 
using OrchardCore.Localization;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Microsoft.Extensions.FileProviders;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.FileProviders.Physical;

public class CustomLocalizationFileLocationProvider : ILocalizationFileLocationProvider
{
    private readonly string _resourcesPath;
    private readonly ILogger<CustomLocalizationFileLocationProvider> _logger;

    public CustomLocalizationFileLocationProvider(string resourcesPath, ILogger<CustomLocalizationFileLocationProvider> logger)
    {
        _resourcesPath = resourcesPath;
        _logger = logger;
        _logger.LogInformation("CustomLocalizationFileLocationProvider instantiated");
    }

    public IEnumerable<IFileInfo> GetLocations(string cultureName)
    {
        _logger.LogInformation($"Getting locations for culture: {cultureName}");
        var cultureFolder = Path.Combine(_resourcesPath, cultureName);
        if (Directory.Exists(cultureFolder))
        {
            var files = Directory.GetFiles(cultureFolder, "*.po").Select(f => new PhysicalFileInfo(new FileInfo(f)));
            _logger.LogInformation($"Found {files.Count()} files for culture '{cultureName}' in '{cultureFolder}'");
            return files;
        }

        _logger.LogWarning($"Culture folder '{cultureFolder}' not found.");
        return Enumerable.Empty<IFileInfo>();
    }
}

Anything else to be done?

**Reply by hishamco** [reply]
I advise you to have a look into `ContentRootPoFileLocationProvider` so you can align your path to be relative to the project root

https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.Localization.Core/PortableObject/ContentRootPoFileLocationProvider.cs

**Reply by VenkatCharyBhairoju** [reply]
Don't we have an option to store PO files in an absolute path and read from there?

**Reply by hishamco** [reply]
The above implementation is quite similar except it reads the PO from Content Root

**Reply by VenkatCharyBhairoju** [reply]
I have created a custom location provider which implements ILocalizationFileLocationProvider as below

`using Microsoft.Extensions.FileProviders;
using Microsoft.Extensions.FileProviders.Physical;
using Microsoft.Extensions.Logging;
using OrchardCore.Localization;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;

public class CustomAbsolutePathPoFileLocationProvider : ILocalizationFileLocationProvider
{
    private readonly string _absolutePath;
    private readonly ILogger<CustomAbsolutePathPoFileLocationProvider> _logger;

    public CustomAbsolutePathPoFileLocationProvider(string absolutePath, ILogger<CustomAbsolutePathPoFileLocationProvider> logger)
    {
        _absolutePath = absolutePath ?? throw new ArgumentNullException(nameof(absolutePath));
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
    }

    public IEnumerable<IFileInfo> GetLocations(string cultureName)
    {
        var cultureInfo = new CultureInfo(cultureName);
        var searchPath = Path.Combine(_absolutePath, cultureInfo.Name);

        _logger.LogInformation("Searching for PO files in {Path}", searchPath);

        if (!Directory.Exists(searchPath))
        {
            _logger.LogWarning("Directory not found: {Path}", searchPath);
            yield break;
        }

        foreach (var file in Directory.EnumerateFiles(searchPath, "*.po", SearchOption.TopDirectoryOnly))
        {
            _logger.LogInformation("Found PO file: {FilePath}", file);
            yield return new PhysicalFileInfo(new FileInfo(file));
        }
    }

    public IEnumerable<IFileInfo> GetLocations()
    {
        _logger.LogInformation("Searching for PO files in {Path}", _absolutePath);

        if (!Directory.Exists(_absolutePath))
        {
            _logger.LogWarning("Directory not found: {Path}", _absolutePath);
            yield break;
        }

        foreach (var file in Directory.EnumerateFiles(_absolutePath, "*.po", SearchOption.AllDirectories))
        {
            _logger.LogInformation("Found PO file: {FilePath}", file);
            yield return new PhysicalFileInfo(new FileInfo(file));
        }
    }
}`

then registered in the Program.cs as below

`
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Localization;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using OrchardCore.Localization;
using System.Globalization;
using System.IO;

var builder = WebApplication.CreateBuilder(args);

// Add services to the container.
builder.Services.AddRazorPages();

// Configure localization options
//builder.Services.Configure<LocalizationOptions>(options =>
//{
//    options.ResourcesPath = "Localization";
//});

// Register the custom provider with the absolute path to PO files
var absolutePath = Path.Combine("C:", "LocalizationFiles");
builder.Services.AddSingleton<ILocalizationFileLocationProvider>(serviceProvider =>
{
    var logger = serviceProvider.GetRequiredService<ILogger<CustomAbsolutePathPoFileLocationProvider>>();
    return new CustomAbsolutePathPoFileLocationProvider(absolutePath, logger);
});

builder.Services.AddPortableObjectLocalization();

builder.Services.Configure<RequestLocalizationOptions>(options =>
{
    var supportedCultures = new[]
    {
        new CultureInfo("en-US"),
        new CultureInfo("fr-FR")
    };
    options.DefaultRequestCulture = new Microsoft.AspNetCore.Localization.RequestCulture("fr-FR");
    options.SupportedCultures = supportedCultures;
    options.SupportedUICultures = supportedCultures;

    options.RequestCultureProviders.Clear();
});

var app = builder.Build();

// Configure the HTTP request pipeline.
if (!app.Environment.IsDevelopment())
{
    app.UseExceptionHandler("/Error");
    app.UseHsts();
}

app.UseHttpsRedirection();
app.UseStaticFiles();
app.UseRouting();

var localizationOptions = app.Services.GetRequiredService<IOptions<RequestLocalizationOptions>>().Value;
app.UseRequestLocalization(localizationOptions);

app.UseAuthorization();

app.MapRazorPages();

app.Run();`

still its not locating PO files from C:/LocalizationFiles folder. Could you please suggest/guide me where it went wrong? Can you please suggest me a link if there is any working samples for this approach. Thanks hishamco!!!

**Reply by hishamco** [reply]
@VenkatCharyBhairoju please create a minimal repo, then share it with me

**Reply by VenkatCharyBhairoju** [reply]
Please find the repo location here. Thanks hishamco!!!

https://github.com/VenkatCharyBhairoju/Localization

Note: Please copy the "LocalizationFiles" to C drive as it have used that as absolute path for po files.

**Reply by VenkatCharyBhairoju** [reply, marked as answer]
Hey @hishamco It started working when I moved the AddPortableObjectLocalization() registration above the custom file provider.


**builder.Services.AddPortableObjectLocalization();**

// Register the custom provider with the absolute path to PO files
var absolutePath = Path.Combine("C:", "LocalizationFiles");
builder.Services.AddSingleton(serviceProvider =>
{
var logger = serviceProvider.GetRequiredService<ILogger>();
return new CustomAbsolutePathPoFileLocationProvider(absolutePath, logger);
});

Thank you so much for your time. I am marking it as Answered

**Reply by hishamco** [reply]
Sorry, @VenkatCharyBhairoju I was OFF the last 3 days, I'm not sure why you need to register `AddPortableObjectLocalization()` again.
